src/domain/flow-system/services/src/flow/flow_service_impl.rs

```diff
@@ -919,6 +929,53 @@ impl FlowService for FlowServiceImpl {
         })
     }
 
+    /// Returns states of flows associated with a given account
+    /// ordered by creation time from newest to oldest
+    /// Applies specified filters
+    #[tracing::instrument(level = "debug", skip_all, fields(%account_name, ?filters, ?pagination))]
+    async fn list_all_flows_by_account(
+        &self,
+        account_name: &AccountName,
+        filters: AccountFlowFilters,
+        pagination: FlowPaginationOpts,
+    ) -> Result<FlowStateListing, ListFlowsByDatasetError> {
+        let account_datasets_ids: Vec<_> = self
+            .dataset_repo
+            .get_datasets_by_owner(account_name)
+            .filter_map_ok(|dataset_handle| {
+                if let Some(dataset_name_filter) = &filters.by_dataset_name {
+                    if dataset_name_filter == &dataset_handle.alias.dataset_name {
+                        return Some(dataset_handle.id);
+                    }
+                    return None;
+                }
+                Some(dataset_handle.id)
+            })
```

Comment:
Try to attach filter only if it is needed. You could split the stream definition into multiple statements using intermediate variables and simple `if` statement.

Reply:
Changed